async.el

```diff
@@ -197,11 +197,29 @@ It is intended to be used as follows:
                              (process-name proc) (process-exit-status proc))))
           (set (make-local-variable 'async-callback-value-set) t))))))
 
-(defun async-read-from-client (proc string)
-  ;; log the message in the process buffer
-  (with-current-buffer (process-buffer proc)
-    (insert string))
+(defun async--process-filter-line-buffering-decorator (filter)
+  "Decorate a process FILTER with line-buffering logic.
+
+Return a new process filter based on FILTER.
+
+FILTER is a function which could be used with `set-process-filter'.
 
+This decorator buffers input until it can pass a complete line
+further to the supplied FILTER.  It is useful as a buffer between
+a process producing data and an Emacs function operating on the
+data which expects to get complete lines as input."
+  (let ((data ""))
+    (lambda (process string)
+      (with-current-buffer (process-buffer process)
+        (insert string))
+
+      (let* ((line-data (split-string (concat data string) "\n")))
```

Comment:
Can't we avoid all this consing and use some navigation or predicate functions instead? E.g.
- by comparing `(line-end-position)` with `(line-end-position 2)`; or
- checking `(char-after (line-end-position))`; or
- using `count-lines`; or
- using `eobp`; or...

Reply:
Well, you need to also remember that:

- 0 new lines can come
- 1 new line can come
- many new lines can come.

The current code is simple (to read) and handles all these cases.  For sure it could be made more efficient.  For example, as @thierryvolpiatto said, we can keep one marker for the "last processed chunk" and then compare end of buffer against that, and when we find a newline inbetween, advance the marker line by line until there is no more newlines.

I think I will rewrite it like that, it sounds that it might be more efficient.